File: samples/scoping-samples/simple-stack-example-scoping-java/src/main/java/com/zhuinden/simplestackexamplescoping/features/words/WordListKey.java
package com.zhuinden.simplestackexamplescoping.features.words;

import com.google.auto.value.AutoValue;
import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.ServiceBinder;
import com.zhuinden.simplestackexamplescoping.core.navigation.BaseKey;

import javax.annotation.Nonnull;

import androidx.fragment.app.Fragment;

@AutoValue
public abstract class WordListKey
        extends BaseKey {
    public static WordListKey create() {
        return new AutoValue_WordListKey();
    }

    @Nonnull
    @Override
    protected Fragment instantiateFragment() {
        return new WordListFragment();
    }

    @Override
    public void bindServices(@Nonnull ServiceBinder serviceBinder) {
        final Backstack backstack = serviceBinder.getBackstack();

        WordController wordController = new WordController(backstack);

        serviceBinder.addService(WordController.class.getName(), wordController);
        serviceBinder.addAlias(WordListFragment.DataProvider.class.getName(), wordController);
        serviceBinder.addAlias(WordListFragment.ActionHandler.class.getName(), wordController);
        serviceBinder.addAlias(NewWordFragment.ActionHandler.class.getName(), wordController);
        serviceBinder.addService("WordEventEmitter", wordController.getEventEmitter());
    }
}

File: samples/scoping-samples/simple-stack-example-scoping-kotlin/src/main/java/com/zhuinden/simplestackexamplescoping/features/words/WordListKey.kt
package com.zhuinden.simplestackexamplescoping.features.words

import com.zhuinden.simplestack.ServiceBinder
import com.zhuinden.simplestackexamplescoping.core.navigation.BaseKey
import com.zhuinden.simplestackextensions.servicesktx.add
import com.zhuinden.simplestackextensions.servicesktx.get
import com.zhuinden.simplestackextensions.servicesktx.rebind
import kotlinx.parcelize.Parcelize

/**
 * Created by Zhuinden on 2018.09.17.
 */
@Parcelize
data object WordListKey : BaseKey() {
    override fun bindServices(serviceBinder: ServiceBinder) {
        with(serviceBinder) {
            add(WordController(backstack))
            rebind<WordListFragment.DataProvider>(get<WordController>())
            rebind<WordListFragment.ActionHandler>(get<WordController>())
            rebind<NewWordFragment.ActionHandler>(get<WordController>())
            add(get<WordController>().eventEmitter)
        }
    }

    override fun instantiateFragment() = WordListFragment()
}